cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -569,24 +574,25 @@ public void removeApplication(final ApplicationId appId) throws Exception {
    * @throws CannotBeDeletedException : the application cannot be deleted because of running programs
    */
   private void ensureNoRunningPrograms(final ApplicationId appId) throws CannotBeDeletedException {
-    //Check if all are stopped.
-    Iterable<ProgramRuntimeService.RuntimeInfo> runtimeInfos =
-      Iterables.filter(runtimeService.listAll(ProgramType.values()),
-                       new com.google.common.base.Predicate<ProgramRuntimeService.RuntimeInfo>() {
-                         @Override
-                         public boolean apply(ProgramRuntimeService.RuntimeInfo runtimeInfo) {
-                           return runtimeInfo.getProgramId().getParent().equals(appId);
-                         }
-                       });
-
-    Set<String> runningPrograms = new HashSet<>();
-    for (ProgramRuntimeService.RuntimeInfo runtimeInfo : runtimeInfos) {
-      runningPrograms.add(runtimeInfo.getProgramId().getProgram());
-    }
-
-    if (!runningPrograms.isEmpty()) {
+    Map<ProgramRunId, RunRecordMeta> runningPrograms = store.getRuns(ProgramRunStatus.RUNNING, null);
```

Comment:
Take a look at `MetadataStoreDataset#getProgramKeyBuilder` methods. Currently, they build a key with either just the recordType or the recordType+programId, but its also possible to build a partial key that is the recordType+appId.
This partial key will only then scan records for that particular appId.